### ISSUE TITLE
Instant-Add API crashes AnkiDroid when AnkiDroid is not running in the background on Android N.

### ISSUE THREAD

**Issue by mmjang** [ankidroid/Anki-Android#4617]
###### Background
I've recently written an app that uses the AnkiDroid Instant Add API. Many users reported that my app crashes when attempting to connect to AnkiDroid (actaully Ankidroid crashes at the same time, and this is the major issue), and their issues can be solved by opening Ankidroid before using my app. After some debuging, I believed that it is either a problem with the `com.ichi2.anki:api:1.1.0alpha5` package or the content provider of Ankidroid. I tried the "API sample" app from the Ankidroid API documentation and the problem remained. 

###### Research

[ x ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid

[ x ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue

[ x ] I have searched for similar existing issues here and on the user forum

###### Reproduction Steps

1. On Android N (I use the official AVD rom from google on windows), install the lasted version of AnkiDroid and the"API sample" app for testing Instant API.
2. Go to settings and stop Ankidroid process to make sure it's not running on the background.
3. Open API sample app, allow requested permission, select an item and share it with the menu item "Ankidroid Instant-Add".
###### Expected Result
Card added to Ankidroid.

###### Actual Result
AnkiDroid crashes.
"API sample" crashes (because the api returned `null`).

###### Debug info
I debuged this on Android Studio and this was the log when crash happened.

###### from api sample app:
```
04-09 04:22:55.597 30484-30484/com.ichi2.apisample E/AndroidRuntime: FATAL EXCEPTION: main
                                                                     Process: com.ichi2.apisample, PID: 30484
                                                                     java.lang.NullPointerException: Attempt to invoke interface method 'java.util.Set java.util.Map.entrySet()' on a null object reference
                                                                         at com.ichi2.apisample.AnkiDroidHelper.getDeckId(AnkiDroidHelper.java:190)
                                                                         at com.ichi2.apisample.AnkiDroidHelper.findDeckIdByName(AnkiDroidHelper.java:167)
                                                                         at com.ichi2.apisample.MainActivity.getDeckId(MainActivity.java:249)
                                                                         at com.ichi2.apisample.MainActivity.addCardsToAnkiDroid(MainActivity.java:272)
                                                                         at com.ichi2.apisample.MainActivity.access$300(MainActivity.java:36)
                                                                         at com.ichi2.apisample.MainActivity$AnkiDroidActionProvider.onMenuItemClick(MainActivity.java:223)
                                                                         at android.support.v7.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:148)
                                                                         at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:958)
                                                                         at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:948)
                                                                         at android.support.v7.view.menu.MenuPopupHelper.onItemClick(MenuPopupHelper.java:191)
                                                                         at android.widget.AdapterView.performItemClick(AdapterView.java:310)
                                                                         at android.widget.AbsListView.performItemClick(AbsListView.java:1156)
                                                                         at android.widget.AbsListView$PerformClick.run(AbsListView.java:3121)
                                                                         at android.widget.AbsListView.onTouchUp(AbsListView.java:4048)
                                                                         at android.widget.AbsListView.onTouchEvent(AbsListView.java:3807)
                                                                         at android.support.v7.widget.ListViewCompat.onTouchEvent(ListViewCompat.java:120)
                                                                         at android.view.View.dispatchTouchEvent(View.java:10023)
                                                                         at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2626)
                                                                         at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2307)
                                                                         at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
                                                                         at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
                                                                         at android.view.ViewGroup.dispatchTransformedTouchEvent(ViewGroup.java:2632)
                                                                         at android.view.ViewGroup.dispatchTouchEvent(ViewGroup.java:2321)
                                                                         at android.widget.PopupWindow$PopupDecorView.dispatchTouchEvent(PopupWindow.java:2275)
                                                                         at android.view.View.dispatchPointerEvent(View.java:10243)
                                                                         at android.view.ViewRootImpl$ViewPostImeInputStage.processPointerEvent(ViewRootImpl.java:4438)
                                                                         at android.view.ViewRootImpl$ViewPostImeInputStage.onProcess(ViewRootImpl.java:4306)
                                                                         at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3853)
                                                                         at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3906)
                                                                         at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3872)
                                                                         at android.view.ViewRootImpl$AsyncInputStage.forward(ViewRootImpl.java:3999)
                                                                         at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:3880)
                                                                         at android.view.ViewRootImpl$AsyncInputStage.apply(ViewRootImpl.java:4056)
                                                                         at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3853)
                                                                         at android.view.ViewRootImpl$InputStage.onDeliverToNext(ViewRootImpl.java:3906)
                                                                         at android.view.ViewRootImpl$InputStage.forward(ViewRootImpl.java:3872)
                                                                         at android.view.ViewRootImpl$InputStage.apply(ViewRootImpl.java:3880)
                                                                         at android.view.ViewRootImpl$InputStage.deliver(ViewRootImpl.java:3853)
                                                                         at android.view.ViewRootImpl.deliverInputEvent(ViewRootImpl.java:6246)
                                                                         at android.view.ViewRootImpl.doProcessInputEvents(ViewRootImpl.java:6220)
                                                                         at android.view.ViewRootImpl.enqueueInputEvent(ViewRootImpl.java:6181)
                                                                         at android.view.ViewRootImpl$WindowInputEventReceiver.onInputEvent(ViewRootImpl.java:6349)
                                                                         at android.view.InputEventReceiver.dispatchInputEvent(InputEventReceiver.java:185)
                                                                         at android.os.MessageQueue.nativePollOnce(Native Method)
                                                                         at android.os.MessageQueue.next(MessageQueue.java:323)
                                                                         at android.os.Looper.loop(Looper.java:136)
                                                                         at android.app.ActivityThread.main(ActivityThread.java:6119)
                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```

###### from AnkiDroid:
```
04-09 04:22:54.970 3088-3088/com.ichi2.anki D/ACRA: ACRA is enabled for com.ichi2.anki, initializing...
04-09 04:22:54.970 3088-3088/com.ichi2.anki D/ACRA: Using custom Report Fields
04-09 04:22:54.974 3088-3088/com.ichi2.anki D/ACRA: Looking for error files in /data/user/0/com.ichi2.anki/files
04-09 04:22:54.975 3088-3088/com.ichi2.anki D/ACRA: Looking for error files in /data/user/0/com.ichi2.anki/files
04-09 04:22:55.002 3088-3100/com.ichi2.anki A/libc: Fatal signal 11 (SIGSEGV), code 2, fault addr 0x8f22a000 in tid 3100 (Binder:3088_1)
                                                    
                                                    [ 04-09 04:22:55.002  1252: 1252 W/         ]
                                                    debuggerd: handling request: pid=3088 uid=10072 gid=10072 tid=3100
04-09 04:22:55.009 3088-3088/com.ichi2.anki W/System: ClassLoader referenced unknown path: /system/app/webview/lib/x86
04-09 04:22:55.066 3105-3105/? A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
04-09 04:22:55.066 3105-3105/? A/DEBUG: Build fingerprint: 'Android/sdk_google_phone_x86/generic_x86:7.1.1/NYC/3756122:userdebug/test-keys'
04-09 04:22:55.066 3105-3105/? A/DEBUG: Revision: '0'
04-09 04:22:55.066 3105-3105/? A/DEBUG: ABI: 'x86'
04-09 04:22:55.066 3105-3105/? A/DEBUG: pid: 3088, tid: 3100, name: Binder:3088_1  >>> com.ichi2.anki <<<
04-09 04:22:55.066 3105-3105/? A/DEBUG: signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0x8f22a000
04-09 04:22:55.066 3105-3105/? A/DEBUG:     eax 00022b68  ebx 8e480333  ecx 0000006e  edx 0000006e
04-09 04:22:55.067 3105-3105/? A/DEBUG:     esi 000a0ba5  edi 8f1e4930
04-09 04:22:55.067 3105-3105/? A/DEBUG:     xcs 00000073  xds 0000007b  xes 0000007b  xfs 0000003b  xss 0000007b
04-09 04:22:55.067 3105-3105/? A/DEBUG:     eip 8e6e6fc2  ebp 00022b78  esp 8f1e4904  flags 00010202
04-09 04:22:55.067 3105-3105/? A/DEBUG: backtrace:
04-09 04:22:55.067 3105-3105/? A/DEBUG:     #00 pc 00010fc2  /data/app/com.ichi2.anki-2/lib/x86/libsqlite3x.so (_ZN7android19utf8ToJavaCharArrayEPKcPti+386)
04-09 04:22:55.067 3105-3105/? A/DEBUG:     #01 pc 0000ed70  /data/app/com.ichi2.anki-2/lib/x86/libsqlite3x.so
04-09 04:22:55.067 3105-3105/? A/DEBUG:     #02 pc 005311b2  /data/app/com.ichi2.anki-2/oat/x86/base.odex (offset 0x4c0000)

```

**Comment by TheKiteRunning** [comment]
This same issue occurs to me as well. 

ANKI can be more efficient with two ancillary apps，like `Android crossword helper` mentioned upstairs and [anki-dict-helper](https://github.com/ninja33/anki-dict-helper) for PC in the way that users can relieve the burden of card making process without loss of the sentence context corresponding to an individual word. 

However,  both of the proposed apps and the reflection process can function better if the instant-add api of ANKI can be more open and lenient. 

So, I really hope you can further consider this issue and forgive my grammar mistakes.

**Comment by onezhou** [comment]
hope u guys fix this problem soon

**Comment by timrae** [comment]
I'm a little tied up with work this month, but I'll take a look when I get a chance. Please feel free to ping me if you don't hear anything back as I may forget.

**Comment by timrae** [comment]
I can't reproduce this on my Nexus 5x (Android 7.1.2) and [v1.0.1](https://github.com/ankidroid/apisample/releases/tag/v1.0.1) of the apisample app which I just built.

Does it only happen on the simulator or on an actual device as well?

**Comment by timrae** [comment]
Also, does the issue happen only with your specific collection, or does it happen with a brand new fresh collection as well?

**Comment by mmjang** [comment]
@timrae Hi Tim, thanks for your reply. The problem is that I can't reproduce this issue on my real device(android 7.1.1) either. But many user of my app report this issue, especially Android N user. They said "the api will not working(crash my app and ankidroid) if ankidroid not running in the background". Maybe I'd better ask for their step by step operations. 

Until now, the only way I can reproduce this issue  is on a simulator. And, yes, on a simulator, this issue happens with an empty collection as well. 

I'm fairly new to Android development, so not sure if I'm missing something or not. Thanks for your work again!

**Comment by timrae** [comment]
It reproducible every time on the simulator?

**Comment by mmjang** [comment]
@timrae recently I added a crash report service to my app, so I can see the crash logs. When add custom models, some users got crashes:
    `#12 java.lang.IllegalArgumentException`
    `Unknown URL content://com.ichi2.anki.flashcards/models`
    `com.ichi2.anki.api.AddContentApi.addNewCustomModel(AddContentApi.java:337)`

One of the user reported that the add cards function are ok, but his was unable to add predefined model to ankidroid from my app. I add pretty sure that I didn't misused the `addNewCustomModel` method, as it works pretty good on my device and emulator. 

I was wandering if the instant add api was somewhat unstable, is it safe for me to connect to the `collection.anki2` sqlite3 databse directly (I will use the query method in the api package)?

**Comment by timrae** [comment]
Are you [checking that the API is available](https://github.com/ankidroid/apisample/blob/master/app/src/main/java/com/ichi2/apisample/AnkiDroidHelper.java#L45) before attempting to use it? Without some code to reproduce the issue there isn't much I can do to help you sorry.

> Is it safe for me to connect to the collection.anki2 sqlite3 databse directly (I will use the query/insert/update method in the api package)?

I strongly recommend you not to do that. 

1. Users are able to customize the path of the database, so you have no way to know where it is

2. You would require your app to have the `WRITE_EXTERNAL_STORAGE ` permission which is not recommended

3. Your app will break anytime we change the db schema

**Comment by mmjang** [comment]
@timrae thanks for your quick replay, I'm pretty sure that the ankidroid database permission is granted before any anki related operations. I even force the user to the permission settings if the default popup window for permission don't show up or the user denied it:

https://github.com/mmjang/ankihelper/blob/890abaeebcc03a5324d9e29aad1c7f6210775a19/app/src/main/java/com/mmjang/ankihelper/ui/LauncherActivity.java#L25

I'll let you know if I find any way to definitely reproduce this issue.

**Comment by jkennethcarino** [comment]
I am able to reproduce this issue on my OnePlus 3 (Android 7.1.1, OxygenOS 4.1.6), AnkiDroid 2.9alpha7, and apisample project (updated to 1.1.0alpha5).

When the apisample calls [`AddContentApi#getDeckList()`](https://github.com/ankidroid/Anki-Android/blob/1c806302ba0c342c34d7996625104ff53e172818/api/src/main/java/com/ichi2/anki/api/AddContentApi.java#L502), the resolver's [query method](https://github.com/ankidroid/Anki-Android/blob/1c806302ba0c342c34d7996625104ff53e172818/api/src/main/java/com/ichi2/anki/api/AddContentApi.java#L504) returns `null`, which causes the app to crash. Also, I am currently working on a plug-in that uses [`AddContentApi#getModelList()`](https://github.com/ankidroid/Anki-Android/blob/1c806302ba0c342c34d7996625104ff53e172818/api/src/main/java/com/ichi2/anki/api/AddContentApi.java#L411) and [`AddContentApi#getDeckList()`](https://github.com/ankidroid/Anki-Android/blob/1c806302ba0c342c34d7996625104ff53e172818/api/src/main/java/com/ichi2/anki/api/AddContentApi.java#L502) to list all available note types and decks, both methods return `null`, but work well when AnkiDroid is running in the background.

**Comment by timrae** [comment]
@mmjang 
>  I'm pretty sure that the ankidroid database permission is granted before any anki related operations. 

That isn't enough, as there is a setting that the user can enable inside AnkiDroid to disable the API entirely. Please additionally check the return value from `getAnkiDroidPackageName()` .

@jkennethcarino

This is a separate issue from the one that @mmjang is reporting, but I tried to reproduce it and couldn't with the latest [AnkiDroid](https://github.com/ankidroid/Anki-Android/releases/latest) and [Apisample](https://github.com/ankidroid/apisample/releases/latest). Can you reproduce on a standard ROM?

**Comment by jkennethcarino** [comment]
@timrae,
I'm unable to reproduce it on my Sony Xperia T2 Ultra (Android 5.1.1) and Wave 8 Tablet (Android 4.4.4).

BTW, here's the first 2 lines of the crash log of apisample on my OnePlus 3:
```
08-21 16:16:01.888 2874-3401/? E/OnePlusAppBootManager: forbid start cpi=ContentProviderInfo{name=com.ichi2.anki.flashcards className=com.ichi2.anki.provider.CardContentProvider}
08-21 16:16:01.888 13524-13524/com.ichi2.apisample E/ActivityThread: Failed to find provider info for com.ichi2.anki.flashcards
```
and the rest of the log:
```
08-21 16:16:01.891 13524-13524/com.ichi2.apisample E/AndroidRuntime: FATAL EXCEPTION: main
                                                                     Process: com.ichi2.apisample, PID: 13524
                                                                     java.lang.NullPointerException: Attempt to invoke interface method 'java.util.Set java.util.Map.entrySet()' on a null object reference
                                                                         at com.ichi2.apisample.AnkiDroidHelper.getDeckId(AnkiDroidHelper.java:190)
                                                                         at com.ichi2.apisample.AnkiDroidHelper.findDeckIdByName(AnkiDroidHelper.java:167)
                                                                         at com.ichi2.apisample.MainActivity.getDeckId(MainActivity.java:249)
                                                                         at com.ichi2.apisample.MainActivity.addCardsToAnkiDroid(MainActivity.java:272)
                                                                         at com.ichi2.apisample.MainActivity.access$300(MainActivity.java:36)
                                                                         at com.ichi2.apisample.MainActivity$AnkiDroidActionProvider.onMenuItemClick(MainActivity.java:223)
                                                                         at android.support.v7.view.menu.MenuItemImpl.invoke(MenuItemImpl.java:167)
                                                                         at android.support.v7.view.menu.MenuBuilder.performItemAction(MenuBuilder.java:973)
                                                                         at android.support.v7.view.menu.MenuPopup.onItemClick(MenuPopup.java:127)
                                                                         at android.widget.AdapterView.performItemClick(AdapterView.java:310)
                                                                         at android.widget.AbsListView.performItemClick(AbsListView.java:1165)
                                                                         at android.widget.AbsListView$PerformClick.run(AbsListView.java:3137)
                                                                         at android.widget.AbsListView$3.run(AbsListView.java:4052)
                                                                         at android.os.Handler.handleCallback(Handler.java:751)
                                                                         at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                         at android.os.Looper.loop(Looper.java:154)
                                                                         at android.app.ActivityThread.main(ActivityThread.java:6290)
                                                                         at java.lang.reflect.Method.invoke(Native Method)
                                                                         at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
                                                                         at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)

```

**Comment by jkennethcarino** [comment]
It seems that the **App auto-launch** feature of OnePlus (Settings > Apps > Gear icon > Auto auto-launch) causes the issue. Either turning off the App auto-launch or allowing AnkiDroid app to the list of auto-launch apps fixes the issue.

**Comment by timrae** [comment]
Good work. Would you be able to make a pull request to prevent the crash in this situation?

**Comment by xyczero** [comment]
@jkennethcarino  I can't find the auto-launch in the step of Gear icon, can you screenshot? thx

**Comment by liowmark** [comment]
https://www.youtube.com/watch?v=THbioOBFhfg

**Comment by liowmark** [comment]
https://www.youtube.com/watch?v=THbioOBFhfg @xyczero

**Comment by mikehardy** [comment]
Fascinating. So, OxygenOS is the vanilla Android on OnePlus devices (referencing an early @timrae comment), and they are the only manufacturer with this "feature", so in order to detect it so that we could warn users that they needed to take action (either disable the feature, or add AnkIDroid to the allowed list) we would have to add a workaround for OxygenOS specifically.

Further, it seems like they have gotten rid of it with their Oreo upgrade, so this may only affect OOS users on N, not the OOS O users. I have a OnePlus 5 personally so I'm familiar with all this stuff, but I've switched to LineageOS so I can't personally help.

Question for anyone: have you upgraded to the Oreo / OOS 5 builds on your OnePlus and if so does it still happen?

Further question: Does this happen to anyone on non-OnePlus devices? I saw a reference to something sort of similar for Xiomi devices but I don't know how many other manufacturers restrict app startup or if it is just OnePlus

Final question: Let's assume that this is only OnePlus OOS 4.x users (just for now), IF this is the case, is it worth it to put a workaround in ankiapi warning users? or is this something users of ankiapi should do? If there would be a pull request, seems like it would be along those lines...

Either way, cool digging @jkennethcarino - following your comment I was able to research this and read up on what's happening

https://stackoverflow.com/questions/45348944/how-to-enable-auto-launch-of-android-application-programaticaly-for-oneplus3-dev  (OnePlus doesn't respond to questions how to workaround it, but others see it)

https://forums.oneplus.net/threads/apps-auto-launch-in-android-n.484674/page-2 (users indicate it is not in OOS 5 builds)

**Comment by mikehardy** [comment]
This is waiting on a response but I read through it again, and I think all we can do (since manufacturers can deny Anki's startup any time) is to alter the API-using apps to detect null and message appropriately that it looks like Anki startup was prevented and cards can't be added until settings are changed. A PR for the apisample that did this is probably the best achievable fix

**Comment by mikehardy** [comment]
I implemented an example check that verifies the API is working prior to attempting to use it. If your app is experiencing this behavior you can [look at it](https://github.com/ankidroid/apisample/pull/5) - specifically [this commit](https://github.com/ankidroid/apisample/pull/5/commits/206b8a82804c20dac9791d865e6a16cf8d66657b#diff-377bc3a3280aa998c1581fc1fed5c3c6) shows how to guard against the API being available but not functioning

if @timrae merges that PR then I think this is closable

**Comment by mikehardy** [comment]
@timrae  this one is done done done (I think)
